bluepysnap/nodes.py

```diff
@@ -239,9 +200,105 @@ def _check_ids(self, node_ids):
 
     def _check_property(self, prop):
+        """Check if a property exists inside the dataset."""
         if prop not in self.property_names:
             raise BluepySnapError("No such property: '%s'" % prop)
 
+    def _get_node_set(self, node_set_name):
+        """Returns the node set named 'node_set_name'."""
+        if node_set_name not in self._node_sets:
+            raise BluepySnapError("Undefined node set: '%s'" % node_set_name)
+        return self._node_sets[node_set_name]
+
+    def _positional_mask(self, node_ids):
+        """Positional mask for the node IDs.
+
+        Args:
+            node_ids (None/numpy.ndarray): the ids array. If None all ids are selected.
+
+        Examples:
+            if the data set contains 5 nodes:
+            _positional_mask([0,2]) --> [True, False, True, False, False]
+        """
+        if node_ids is None:
+            return np.full(len(self._data), fill_value=True)
+        mask = np.full(len(self._data), fill_value=False)
+        mask[node_ids] = True
+        return mask
+
+    def _node_population_mask(self, queries):
+        """Handle the population and node ID queries."""
+        populations = queries.pop(POPULATION_KEY, None)
+        if populations is not None and self.name not in set(utils.ensure_list(populations)):
+            node_ids = []
+        else:
+            node_ids = queries.pop(NODE_ID_KEY, None)
+        return queries, self._positional_mask(node_ids)
+
+    def _properties_mask(self, queries):
+        """Return mask of node IDs with rows matching `props` dict."""
+        # pylint: disable=assignment-from-no-return
+        unknown_props = set(queries) - set(self._data.columns) - {POPULATION_KEY, NODE_ID_KEY}
+        if unknown_props:
+            raise BluepySnapError("Unknown node properties: [{0}]".format(", ".join(unknown_props)))
+
+        queries, mask = self._node_population_mask(queries)
+        if not mask.any():
+            # Avoid fail and/or processing time if wrong population or no nodes
+            return mask
+
+        for prop, values in six.iteritems(queries):
+            prop = self._data[prop]
+            if issubclass(prop.dtype.type, np.floating):
+                v1, v2 = values
+                prop_mask = np.logical_and(prop >= v1, prop <= v2)
+            elif isinstance(values, six.string_types) and values.startswith('regex:'):
+                prop_mask = _complex_query(prop, {'$regex': values[6:]})
+            elif isinstance(values, collections.Mapping):
+                prop_mask = _complex_query(prop, values)
+            else:
+                prop_mask = np.in1d(prop, values)
+            mask = np.logical_and(mask, prop_mask)
+        return mask
+
+    def _operator_mask(self, queries):
+        """Handle the query operators '$or', '$and'."""
+        # will pop the population and or/and operators so need to copy
+        queries = deepcopy(queries)
```

Comment:
You would not need the copy if you would use
`queries = queries[1:]` instead of the `pop`, no ?

Reply:
then it becomes a destructive operation for queries which I don t want.